### ISSUE TITLE
Add typechecking via Mypy

### ISSUE THREAD

**Issue by mako-taco** [ReactiveX/RxPY#186]
Hello RxPY!
I need to create type stubs for RxPY in order to use RxPY at my work. It's not much more effort to instead add types to the existing library. I would be willing to add [mypy](http://mypy-lang.org/) types to RxPY and and update your travis files to monitor the status of the build if you'd be willing to accept the PR (It's a lot of work to add the types initially, and I'd rather not maintain a fork).

Disclaimer: I currently sit 10ft away from the team that's building mypy

Please let me know your opinion on the matter.

**Comment by thomasnield** [comment]
Feel free to chime in @dbrattli , but I'm a little reluctant to having anything related to `mypy` implemented on the main branch of RxPy, especially since `mypy` is still in development. However, I'm sure it's more than okay to create a fork, and if that is successful we can discuss next steps down the road.

**Comment by ubald** [comment]
They're not strictly mypy types, they are in supported since 3.5 (though provisionally: https://docs.python.org/3/library/typing.html) and for backwards compatibility, can be added in a sidecar file that won't mess up any of RxPY's code.

**Comment by dbrattli** [comment]
Have I understood things correctly that this is about creating stub files and not about adding type annotation to the existing files? In that case I see no problems other than keeping them in sync.

**Comment by graingert** [comment]
FYI you can use comments to add type annotations

**Comment by graingert** [comment]
Then you can use mypy to verify that you're in sync

**Comment by dbrattli** [comment]
Sorry for the delay. I'm planning to merge [Aioreactive](https://github.com/dbrattli/aioreactive) and RxPY into RxPY 2.0. The idea is to take many of the lessons learned from aioreactive into RxPY incl. Python 3 and type annotations for everything. I will setup a new branch shortly for working on this experiment.

**Comment by graingert** [comment]
Nice, will it still have pluggable event loop support?

What about python 2 support?

I'd like to use rxpy to support python 2+twisted/python 3+asyncio in one
codebase

**Comment by dbrattli** [comment]
I'm unsure about Python 2 support. I would like `Observable` and `AsyncObservable` (async/await) in a single codebase with type annotations. It might be that we have to suggest RxPY v1 for Python 2 support.

**Comment by michaeljohnbennett** [comment]
is this Aioreactive stuff integrated into the codebase now @dbrattli? Would be interested to try it out if so.

**Comment by dbrattli** [comment]
No, it will not be integrated. It will remain a separate project but the plan is to align it with RxPY v3 so you can pipe back and forth between Obserables and AsyncObservables.
